abi-bindings/go/packer/packer_test.go

```diff
@@ -0,0 +1,190 @@
+
+// Paths below are relative to the current package.
+const packagePath = "../..."
+const currentPackagePath = "."
+
+var fs = token.NewFileSet()
+
+var packerTypes = map[string]ABIPacker{
+	"ValidatorSetSigMessage": &validatorsetsig.ValidatorSetSigMessage{},
+	"TeleporterMessage":      &teleportermessenger.TeleporterMessage{},
+	"ProtocolRegistryEntry":  &teleporterregistry.ProtocolRegistryEntry{},
+}
+
+func findAllImplementers(t *testing.T) []string {
+	cfg := &packages.Config{
+		Mode: packages.NeedName | packages.NeedFiles | packages.NeedSyntax | packages.NeedTypes | packages.NeedTypesInfo,
+		Fset: fs,
+	}
+	interfacePkg, err := packages.Load(cfg, currentPackagePath)
+	require.NoError(t, err)
+
+	var iface *types.Interface
+	// find the interface in the currnet package
+	for _, pkg := range interfacePkg {
+		for _, obj := range pkg.Types.Scope().Names() {
+			if obj == interfaceName {
+				ifaceObj := pkg.Types.Scope().Lookup(obj)
+				if ifaceObj == nil {
+					continue
+				}
+				if typ, ok := ifaceObj.Type().Underlying().(*types.Interface); ok {
+					iface = typ
+					break
+				}
+			}
+		}
+	}
+	require.NotNilf(t, iface, "Interface %s not found in package %s", interfaceName, currentPackagePath)
+
+	pkgs, err := packages.Load(cfg, packagePath)
+	require.NoError(t, err)
+
+	allImplementers := []string{}
+
+	for _, pkg := range pkgs {
+		for _, file := range pkg.Syntax {
+			// Inspect types in the package
+			for _, decl := range file.Decls {
+				if genDecl, ok := decl.(*ast.GenDecl); ok {
+					for _, spec := range genDecl.Specs {
+						if typeSpec, ok := spec.(*ast.TypeSpec); ok {
+							obj := pkg.Types.Scope().Lookup(typeSpec.Name.Name)
+							if obj == nil {
+								continue
+							}
+							if _, ok := obj.Type().Underlying().(*types.Struct); ok {
+								typeName := typeSpec.Name.Name
+								obj := pkg.Types.Scope().Lookup(typeName)
+								if obj != nil {
+									if named, ok := obj.Type().(*types.Named); ok {
+										if types.Implements(types.NewPointer(named), iface) {
+											allImplementers = append(allImplementers, named.Obj().Name())
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	return allImplementers
+}
+
+func TestExhaustivePacking(t *testing.T) {
+	implementers := findAllImplementers(t)
+	for _, structName := range implementers {
+		packerType, ok := packerTypes[structName]
+		require.True(
+			t,
+			ok,
+			fmt.Sprintf("Struct %s not found in packer_test.go packerTypes map. Add it to check for exhaustiveness", structName),
+		)
+
+		randomizedStruct, err := randomizeStruct(packerType)
+		require.NoError(t, err)
+
+		v1 := randomizedStruct.(ABIPacker)
+		encoded, err := v1.Pack()
+		require.NoError(t, err)
+		v2 := reflect.New(reflect.TypeOf(v1).Elem()).Interface().(ABIPacker)
+		err = v2.Unpack(encoded)
+		require.NoError(t, err)
+		require.Equal(t, v1, v2)
+	}
+}
+
+func randomizeStruct(packerStruct interface{}) (interface{}, error) {
```

Comment:
If we think that these randomization functions are useful we can maybe add them to a test utils location?